pkg/webhook/resources/persistentvolumeclaim/mutator.go

```diff
@@ -47,32 +48,86 @@ func (m *pvcMutator) Create(_ *types.Request, newObj runtime.Object) (types.Patc
 	pvc := newObj.(*corev1.PersistentVolumeClaim)
 
 	logrus.Debugf("create PVC %s/%s with mutator", pvc.Namespace, pvc.Name)
+	patchOPs := []string{}
+	isGoldenImage := false
 
 	// check pvc is related to the vm image
-	_, err := m.vmImageCache.Get(pvc.Namespace, pvc.Name)
+	patchOp, err := m.patchGoldenImageAnnotation(pvc)
 	if err != nil {
+		return nil, err
+	}
+	if patchOp != "" {
+		isGoldenImage = true
+		patchOPs = append(patchOPs, patchOp)
+	}
+
+	// patch populator
+	if !isGoldenImage {
+		patchOp, err = m.patchDataSource(pvc)
+		if err != nil {
+			return nil, err
+		}
+		patchOPs = append(patchOPs, patchOp)
+	}
+
+	// no need to patch
+	if len(patchOPs) == 0 {
+		return nil, nil
+	}
+	return patchOPs, nil
+}
+
+func (m *pvcMutator) patchGoldenImageAnnotation(pvc *corev1.PersistentVolumeClaim) (string, error) {
+	if _, err := m.vmImageCache.Get(pvc.Namespace, pvc.Name); err != nil {
 		if apierrors.IsNotFound(err) {
 			logrus.Infof("PVC %s/%s is not related to the VM image, skip patch", pvc.Namespace, pvc.Name)
-			return nil, nil
+			return "", nil
 		}
-		return nil, err
+		return "", err
 	}
 
 	annotations := pvc.GetAnnotations()
 
-	if _, find := annotations[util.AnnotationGoldenImage]; find {
-		if annotations[util.AnnotationGoldenImage] == "true" {
-			return nil, nil
-		}
+	if v, find := annotations[util.AnnotationGoldenImage]; find && v == "true" {
+		return "", nil
 	}
 	annotations[util.AnnotationGoldenImage] = "true"
 
 	annoVal, err := json.Marshal(annotations)
 	if err != nil {
 		logrus.Warnf("failed to marshal annotations: %v, err: %v", annotations, err)
-		return nil, err
+		return "", err
+	}
+
+	// patch annotation
+	return fmt.Sprintf(`{"op": "replace", "path": "/metadata/annotations", "value": %s}`, string(annoVal)), nil
```

Comment:
Assuming that K8s call all mutators sequentially, this should be race free. Otherwise, we should append our annotation to the end of the list. See example in the [JSON patch RFC](https://datatracker.ietf.org/doc/html/rfc6902#appendix-A.16).

Reply:
I am not sure what you mean.
I get the whole annotation and add (or do no-op) with the current PVC annotation.
So, it should not break any annotation.
Could you explain more about this?

Reply:
Say if another CSI provider has a mutator that also modifies the annotations via JSON patch. If our mutator doesn't see these modified annotations, then when the K8s API Server sees this `replace` JSON patch, it will replace the whole annotations with what we submitted here, potentially, overriding the modified annotations from the other CSI provider. It has been a while since I last looked at K8s mutator invocation code, but I think they are called sequentially, meaning, it _should be_ safe to do a `replace` op here.

Reply:
Thanks for the detailed explanation. I know your concern, and now I only patch the annotation key we wanted.